File: PerlInterviewQuestions/checkPrs.pl
use strict;
use warnings;

sub checkPrs {
    die "Wrong number of args" if (scalar(@_) != 1);
    my $mystr = $_[0];
    my @stack = ();
    my %s = ("{" => undef, "(" => undef, "[" => undef);
    my %d = ("}" => "{", ")" => "(", "]" => "[");
    foreach my $char (split('', $mystr)) {
        if (exists $s{$char}) {
            push(@stack, "$char");
        }
        if (exists($d{$char})) {
            if ($#stack + 1 == 0) {
                return "False";
            }
            else {
                if (exists $d{$char}) {
                    pop(@stack)
                }
            }
        }
    }
    $#stack + 1 == 0 ? return "True" : return "False";
}

print(checkPrs("((())") . "\n");
print(checkPrs("((()))") . "\n");
print(checkPrs("[((()))]") . "\n");
print(checkPrs("}((())") . "\n");

File: PythonInterviewQuestions/checkPrs.py
def checkPrs(mystr):
    s = set(["(", "[", "{"])
    d = {")": "(", "]": "[", "}": "{"}
    stack = []

    for c in mystr:
        if c in s:
            stack.append(c)
            continue

        if c in d:
            if len(stack) == 0:
                return False
            else:
                stack.pop()

    return len(stack) == 0


def main():
    print(checkPrs("((())"))
    print(checkPrs("((()))"))
    print(checkPrs("[((()))]"))
    print(checkPrs("}((())"))


if __name__ == '__main__':
    main()
